src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -305,6 +315,86 @@ export class TerminalService implements ITerminalService {
 
 		// Create async as the class depends on `this`
 		timeout(0).then(() => this._instantiationService.createInstance(TerminalEditorStyle, document.head));
+
+		//TODO:@meganrogge is there a better way to do this?
+		this.onDidReceiveProcessId(async () => await this._refreshDescriptionCwd());
+		this.onDidDisposeInstance(async () => await this._refreshDescriptionCwd());
+	}
+
+	private async _refreshDescriptionCwd(): Promise<void> {
+		if (!this._configHelper.config.description.includes('cwdFolder')) {
+			return;
+		}
+		const folderToTerminals: Map<string, TerminalCwd[]> = new Map();
+		const foldersToDistinguish: string[] = [];
+		for (const terminal of this.instances) {
+			const cwd = await terminal.getCwd();
+			const folder = basename(cwd);
+			const termsForThisFolder = folderToTerminals.get(folder);
+			if (!termsForThisFolder) {
+				folderToTerminals.set(folder, [{ terminal, cwd }]);
+			} else {
+				let shouldDistinguish = termsForThisFolder.find(t => t.cwd !== cwd);
+				if (shouldDistinguish || termsForThisFolder.length === 1) {
+					termsForThisFolder.push({ terminal: terminal, cwd });
+					folderToTerminals.set(folder, termsForThisFolder);
+					if (!foldersToDistinguish.includes(folder)) {
+						foldersToDistinguish.push(folder);
+					}
+				}
+			}
+			for (const folder of foldersToDistinguish) {
+				let terms = folderToTerminals.get(folder);
+				if (!terms) {
+					return;
+				}
+
+				terms = this._differentiateFolders(terms);
+
+				for (const term of terms) {
+					//TODO:@meganrogge is there a better way to do this? maybe a new TitleEventSource?
+					term.terminal.setTitle(term.cwd, TitleEventSource.Api, term.cwd);
+				}
+			}
+		}
+	}
+
+	private _differentiateFolders(terms: TerminalCwd[]): TerminalCwd[] {
+		const uniqueCwds = Array.from(new Set(terms.map(t => t.cwd)));
+		const paths = [];
+		const cwdToPath: Map<string, string[]> = new Map();
+		for (const uniqueCwd of uniqueCwds) {
+			paths.push(uniqueCwd.split(isWindows ? '\\' : '/'));
+			cwdToPath.set(uniqueCwd, uniqueCwd.split(isWindows ? '\\' : '/'));
+		}
+		let index = 1;
+		const minLength = paths.sort((a, b) => a.length - b.length)[0].length;
+		while (new Set(paths.map(p => p[p.length - index])).size !== paths.length && minLength > index) {
```

Comment:
But hard to read and not a fan of instantiating a Set inside a loop. I'm guessing this is copied from the editor code? We can talk about sharing the code when we chat today.

Reply:
no I didn't look at the editor code, just implemented it myself because it's slightly diff't since we need to keep track of the terminal